Add Portfile for Yarn 0.18.1

jambonrose opened the pull request:
Hi,
This PR adds support for installation of [yarn](https://yarnpkg.com/), Facebook's Javascript dependency manager.

The Portfile follows the spirit of the [bash installer found on the yarn website](https://yarnpkg.com/install.sh), but instead of installing yarn in the users' home directory, yarn is placed in Macports' `libexec` directory. This installation is therefore **not global**.

The installation is independent of [NPM](https://www.npmjs.com/) (what yarn aims to replace).

Yarn requires [Node](http://nodejs.org/) to run. I've added Node 6 as default (as it's the current version with Long-Term Support), but have added installation variants to force Node6 via Macports (even if it's already installed), or else to use Node 4 or Node 7. As Node 5 is no longer supported or maintained, I have not included it.

The `reinplace` invocation in the patch phase changes the line of code in 1 into the line of code in 2.

1. `basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")`
2. `basedir=$(dirname "$(readlink "$0" || echo "$(echo "$0" | sed -e 's,\\,/,g')")")`

This enables the script to be soft-linked in `${prefix}/bin` and called from Macports' structure. This step is essential, as it's what allows yarn to be installed in `${prefix}/libexec` instead of globally.

I would prefer to use a regular expression capturing group, but I wasn't able to get that working. I was thinking of something like the code below (where the `\1` prints the code captured by `(.+)`).

```tcl
reinplace -E "s;basedir=\\$\\(dirname (.+)\\);basedir=\$(dirname \"\$(readlink \"\$0\" || echo \1)\");" ${worksrcpath}/bin/yarn
```

Following PR #43, I've not included the line `# $Id$`.

Thanks for reviewing! Feedback welcome.

larryv commented on devel/yarn/Portfile:
```diff
@@ -0,0 +1,52 @@
+
+post-patch {
+    reinplace -E "s;basedir=\\$\\(.+\\);basedir=\$(dirname \"\$(readlink \"\$0\" || echo \"\$(echo \"\$0\" | sed -e 's,\\\\\\\\,/,g')\")\");" ${worksrcpath}/bin/yarn
+}
```
This should be a patch.

larryv commented on devel/yarn/Portfile:
```diff
@@ -0,0 +1,52 @@
+depends_run         bin:node:nodejs6
+variant nodejs4 conflicts nodejs6 nodejs7 description {Use Node.js v4} {
+    depends_lib-replace bin:node:nodejs6 port:nodejs4
+}
```
You use `depends_run` initially but `depends_lib` in the variants. Which is it?

If the original dependency accepts any `node` executable, what is the purpose of the variants?

nerdling replied:
Likely `depends_run`, since it is node code being run. As far as versioning, isn't it the usual MacPorts pattern to include a specific one to ensure everything might run the same version of node? Like editing shebangs, configuring paths to look for libraries, etc.

jambonrose replied:
Yes, I meant to use `depends_run`.

I think @larryv raises a good point - the variants are unnecessary. I will remove them, but I will keep the `depends_run`.

@nerdling, I'm afraid you've lost me on the subject of versioning. Should I be doing things differently?

larryv commented on devel/yarn/Portfile:
```diff
@@ -0,0 +1,52 @@
+destroot {
+  copy ${worksrcpath} ${destroot}${prefix}/libexec/yarn
+  ln -s ${prefix}/libexec/yarn/bin/yarn ${destroot}${prefix}/bin/yarn
+}
```
Why install this into `libexec` instead of directly into `bin`?

jambonrose replied:
Yarn is not a stand-alone executable. It's a shell script that invokes a JavaScript/Node script, which relies on other JavaScript files in the various directories around it.

I opted to put the scripts in `libexec` and to soft-link the shell script. We could put all the scripts and directories in `bin`, but that feels messy, especially because you don't want to call those scripts directly. We could alternatively put the scripts in `lib`, but no other program is going to rely on these scripts, so `libexec` seemed most appropriate.

Note that we could also soft-link the JavaScript script directly (instead of using the shell script) but I feel like I'm circumventing project code at that point.

jambonrose commented:
@larryv @nerdling : Thanks for the review!

I discovered a bit of a hiccup in my install of yarn yesterday: the global command acts in very strange ways. As soon as I figure out what's going on there I will update this PR.

isomarcte commented:
@jambonrose I saw you comment on https://trac.macports.org/ticket/52827 I have combined what I _think_ are the best of both of our ports, as well as the feedback on this PR. You can find it on the attachment to the trac ticket. Assuming you like the combination of our approaches, we can just add them to this PR and close the ticket.

Let me know what you think.

jambonrose commented:
PR Update:

- implemented @larryv's changes
- updated Yarn to 0.18.0 (and changed the title of this PR 1.17.3->0.17.3->0.18.0)
- switched to the github Portgroup
- added notes based on @isomarcte's patch
    
I goofed while writing the initial commit message. This was for version 0.17.3, not 1.17.3. @larryv @nerdling : Would you like me to rebase/change the message and force push? If git history is important, I'm happy to do so.

@isomarcte: Thanks for taking a look at this. A couple reactions to your patch:

- I did not know about livechecks! I think that's a compelling reason to switch to using Github. I've updated the PR as such.

- Node v6 is LTS and the recommended install (per http://nodejs.org/), so I think we should prefer that to Node v7 if Node isn't installed already.

- The patch you created replaces `echo` with `readlink -e`. I'm running the default macOS 10.11 `readlink`, and I don't have an `e` flag. My script achieves the same behavior as the `e`/`f` flag on GNU's `readlink` by using the `|| echo "$0"`. What's more, I've opened a PR on yarnpkg/yarn#2217 to replace the code directly, following two positive pieces of feedback. 

- I understand that it mirrors the NPM structure, but I don't see a compelling reason to create a `node_module` directory to store yarn in. Following the Manual Installation from tarball section on the [installation page](https://yarnpkg.com/en/docs/install#alternatives-tab) feels more appropriate to me.

mojca commented on devel/yarn/Portfile:
```diff
@@ -29,17 +29,8 @@ checksums           rmd160  be20072bfd66a6411911bebb63284abb7f7f454c \
-}
 
-post-patch {
+patch {
```
I believe that you misunderstood what @larryv wanted to tell you. Not that this should be changed from `post-patch` into `patch` (you should probably never do that), but you should create a proper patch file, perhaps call it patch-bin-yarn.diff, put it under `files/` and add a line
```
patchfiles patch-bin-yarn.diff
```
to the Portfile.

Regular expressions are too fragile, in particular when upgrading to a newer versions and this one is definitely too complex. The `reinplace` is only used when you need to substitute `${prefix}` or so.

In fact I believe that this is what you want. I don't want to decrypt or guess what your expression tries to do, but you can replace the whole line with
```
basedir=@@PREFIX@@
```
and then do
```
reinplace "s|@@PREFIX@@|${prefix}|g"
```
I'm not sure if that's actually the proper path, but I guess it might be something in that direction.

mojca commented on devel/yarn/Portfile:
```diff
@@ -0,0 +1,73 @@
+
+depends_run         bin:node:nodejs6
+
+patch {
```
I believe that you misunderstood what @larryv wanted to tell you. Not that this should be changed from `post-patch` into `patch` (you should probably never do that), but you should create a proper patch file, perhaps call it `patch-bin-yarn.diff,` put it under `files/` and add a line
```
patchfiles patch-bin-yarn.diff
```
to the Portfile.

Regular expressions are too fragile, in particular when upgrading to a newer versions and this one is definitely too complex. The `reinplace` is only used when you need to substitute `${prefix}` or so.

In fact I believe that this is what you want. I don't want to decrypt or guess what your expression tries to do, but you can replace the whole line with something similar to
```
basedir=@@PREFIX@@
```
and then do
```
post-patch {
    reinplace "s|@@PREFIX@@|${prefix}|g" ${worksrcpath}/bin/yarn
}
```
I'm not sure if that's actually the proper variable or value, but I guess it might be something in the desired direction.

(I'm sorry, I commented on the wrong place earlier, fixing this now.)

isomarcte commented:
Hey @jambonrose,

> updated Yarn to 0.18.0 (and changed the title of this PR 1.17.3->0.17.3->0.18.0)

I would set this at `0.17.10`. The livecheck correctly reports `0.18.0` as the most recent tarball, but I noticed this morning that the yarn folks have labeled that as a pre-release.

> Node v6 is LTS and the recommended install (per http://nodejs.org/), so I think we should prefer that to Node v7 if Node isn't installed already.

Sounds good to me.

> The patch you created replaces echo with readlink -e. I'm running the default macOS 10.11 readlink, and I don't have an e flag. My script achieves the same behavior as the e/f flag on GNU's readlink by using the || echo "$0". What's more, I've opened a PR on yarnpkg/yarn#2217 to replace the code directly, following two positive pieces of feedback.

Yeah, I noticed this today too. Glad you working on it.

> I understand that it mirrors the NPM structure, but I don't see a compelling reason to create a node_module directory to store yarn in. Following the Manual Installation from tarball section on the installation page feels more appropriate to me.

They are just installing it directly into `/opt` which isn't really applicable to macports. IMHO, we should do the same thing with `yarn` that we do with `npm` since they fill much of the same purpose in the ecosystem.

But I'll you guys decide.

I am going to make a note on the ticket that it should be closed an this PR considered the canonical source.

isomarcte commented:
@jambonrose

I just saw your comment on the ticket about the `distname` and `worksrcdir`.

I think you need to leave them in. The github port group assumes distname is defined like so,

```tcl
distname                ${github.project}-${github.version}
```

And the contents of the tarball is a directory of the same name as the tarball. This _should_ be true, but sadly the yarn guys are tarring a folder named `dist` instead of `yarn-v0.18.0`.

jambonrose commented:
PR Update:

- Use patch/diff file instead of regular expression
- Downgrade to version 0.17.10

@mojca: You're absolutely right; I completely misunderstood @larryv's instructions. Thanks very much for taking the time to explain the issue. I've updated the PR to use a patch file.

@isomarcte: I wasn't sure what was going on with the versions. I think that you're right, though, especially given that the website tarball still uses 0.17.10, and have switched to that version. A big thank you for clarifying `distname` and `worksrcdir`!

For future reference: yarnpkg/yarn#2217 has been merged, meaning future versions of the portfile should remove or replace the existing patch.

I believe this is now ready to be merged.

janvogt commented:
Any news on this? People (i.e. Hombrew users) mock me for not being able to install yarn via MacPorts... ;D

mojca commented:
Looks good to me. But just curious, why did you have to revert from 0.18.0 back to 0.17.10? Meanwhile there is 0.18.1 out (and there is no "prerelease" tag attached to it), I guess that one should be OK? Sadly your patch has only been merged to the master branch, but that's ok.

@larryv: are you ok with the current state of the port?

Of course we need to squash the changes into a single commit.

jambonrose commented:
Hi @mojca, Thanks for looking this over, and for catching that they hadn't merged my PR into the v0.18 branch.

Re: v0.18
I think we were apprehensive of the pre-release tag, but also that the official website was still using v0.17.10 as the version to be downloaded.

Following your comments and the fact that v0.18.1 is on both github and the yarn website, I've upgraded the PR to use that version.
